lib/shim/message-shim.js

```diff
@@ -356,108 +354,37 @@ function recordConsume(nodule, properties, spec) {
-        if (destIdx !== null) {
-          msgDesc.destinationName = args[destIdx]
-        }
-      }
-
-      // Make the segment if we can.
-      if (!msgDesc) {
-        shim.logger.trace('Not recording consume, no message descriptor.')
-        return fn.apply(this, args)
-      }
-
-      const name = _nameMessageSegment(shim, msgDesc, shim._metrics.CONSUME)
-
-      // Adds details needed by createSegment when used with a spec
-      msgDesc.name = name
-      msgDesc.recorder = genericRecorder
-      msgDesc.parent = parent
-
-      const segment = shim.createSegment(msgDesc)
-      const getParams = shim.agent.config.message_tracer.segment_parameters.enabled
-      const resHandler = shim.isFunction(msgDesc.messageHandler) ? msgDesc.messageHandler : null
-
-      const cbIdx = shim.normalizeIndex(args.length, msgDesc.callback)
-      if (cbIdx !== null) {
-        shim.bindCallbackSegment(args, cbIdx, segment)
-
-        // If we have a callback and a results handler, then wrap the callback so
-        // we can call the results handler and get the message properties.
-        if (resHandler) {
-          shim.wrap(args, cbIdx, function wrapCb(shim, cb, cbName) {
-            if (shim.isFunction(cb)) {
-              return function cbWrapper() {
-                const cbArgs = shim.argsToArray.apply(shim, arguments)
-                const msgProps = resHandler.call(this, shim, cb, cbName, cbArgs)
-                if (getParams && msgProps && msgProps.parameters) {
-                  shim.copySegmentParameters(segment, msgProps.parameters)
-                }
-
-                return cb.apply(this, arguments)
-              }
-            }
-          })
-        }
+  return this.record(nodule, properties, function consumeRecorder(shim, fn, fnName, args) {
+    if (shim.isFunction(spec)) {
+      spec = spec.apply(this, arguments)
+    } else {
+      spec = { ...DEFAULT_SPEC, ...spec }
+      const destIdx = shim.normalizeIndex(args.length, spec.destinationName)
+      if (destIdx !== null) {
+        spec.destinationName = args[destIdx]
       }
-
-      // Call the method in the context of our segment.
-      let ret = shim.applySegment(fn, segment, true, this, args)
-
-      if (ret && msgDesc.promise && shim.isPromise(ret)) {
-        ret = shim.bindPromise(ret, segment)
-
-        // Intercept the promise to handle the result.
-        if (resHandler) {
-          ret = ret.then(function interceptValue(res) {
-            const msgProps = resHandler.call(this, shim, fn, fnName, res)
-            if (getParams && msgProps && msgProps.parameters) {
-              shim.copySegmentParameters(segment, msgProps.parameters)
-            }
-            return res
-          })
+    }
+    const messageHandler = spec.messageHandler
+    const getParams = shim.agent.config.message_tracer.segment_parameters.enabled
+
+    if (messageHandler) {
+      spec.after = function afterPromise({ shim, val, segment }) {
```

Comment:
this really cuts down on so much code, nice work 👏🏻

Reply:
this is a breaking change right? we've changed the signature to be object based.  I know we updated our code to now parse the object but customers may be unhappy.  I will take with semver:major and maybe we can release a semver major soon with this, my refactor of registering multiple hooks and #1520

Reply:
So, we do not publicly document what should go into`spec.after`, other than saying that it's a callback function. I guess this is a breaking change but it's such an obscure part of our API that I wonder if anyone even uses it.

However, if we already have another reason to introduce a breaking change, this is a good place to do it.

Reply:
shim API is a grey area because we expose it. Even if we don't document it, it's publicly available. We should just ship this with our next semver major so hold off on merging. We discussed in standup today that we will be making a semver major release shortly(maybe this week)